### PR TITLE
Add typing for steps/discover/__init__.py

### PR DESCRIPTION
**Pull request by janhavlin**

### CODE REVIEW

File: tmt/steps/discover/__init__.py [hunk 1]
```diff
@@ -94,7 +104,7 @@ def wake(self):
 
         # Choose the right plugin and wake it up
         for data in self.data:
-            plugin = DiscoverPlugin.delegate(self, data)
+            plugin = DiscoverPlugin.delegate(self, dict(data))
```
Comment (thrix) [inline tmt/steps/discover/__init__.py:107]: hmm, isn't really type annotation of `delegate` function incorrect?

It is now

```
  445     @classmethod
  446     def delegate(
  447             cls,
  448             step: Step,
  449             data: Dict[Any, Any]) -> 'Plugin':
```

shouldn't `data` be rather have `StepData` type?

File: tmt/steps/discover/__init__.py [hunk 2]
```diff
@@ -107,12 +117,12 @@ def wake(self):
             self.status('todo')
             self.save()
 
-    def show(self):
+    def show(self) -> None:
         """ Show discover details """
         for data in self.data:
-            DiscoverPlugin.delegate(self, data).show()
+            DiscoverPlugin.delegate(self, dict(data)).show()
```
Comment (thrix) [inline tmt/steps/discover/__init__.py:123]: same here ...

File: tmt/steps/discover/__init__.py [hunk 3]
```diff
@@ -265,6 +279,7 @@ def extract_distgit_source(
             stdout, _ = self.run(
                 ["git", "config", "--get-regexp", '^remote\\..*.url'],
                 cwd=distgit_dir)
+            assert stdout is not None
```
Comment (thrix) [inline tmt/steps/discover/__init__.py:282]: @janhavlin I would rather return a resonable error here? As I understand the code, this should be really handled as a reasonable error, it is an error condition to me.

File: tmt/steps/discover/__init__.py (hunk 1 above)
Comment (janhavlin) [reply tmt/steps/discover/__init__.py:107]: Fyi, there's also a discussion about this same line in an another PR: https://github.com/teemtee/tmt/pull/1319#discussion_r916599351

After looking briefly at other `delegate()` method calls, it seems that changing the type to `StepData` would indeed be more suitable.

Comment (psss) [reply tmt/steps/discover/__init__.py:107]: Yes, this should be covered by ffaa8ec303e33156f4517a5dc9bad7f3cf7d11b5.

File: tmt/steps/discover/__init__.py (hunk 3 above)
Comment (janhavlin) [reply tmt/steps/discover/__init__.py]: Changed it to raising a `GeneralError`. Is it okay now?

**Review by optak — APPROVED** [review]
LGTM

File: tmt/steps/discover/__init__.py [hunk 4]
```diff
@@ -1,24 +1,33 @@
...
-from typing import Optional, Type
+from typing import TYPE_CHECKING, Any, List, Optional, Type, cast
 
 import click
 from fmf.utils import listed
 
 import tmt
+from tmt.steps import StepData
+
+if TYPE_CHECKING:
```
Comment (guy9050) [inline tmt/steps/discover/__init__.py:10]: what is this for?

File: tmt/steps/discover/__init__.py [hunk 5]
```diff
@@ -269,6 +279,9 @@ def extract_distgit_source(
             stdout, _ = self.run(
                 ["git", "config", "--get-regexp", '^remote\\..*.url'],
                 cwd=distgit_dir)
+            if stdout is None:
+                raise tmt.utils.GeneralError("Missing remote origin url.")
+
```
Comment (guy9050) [inline tmt/steps/discover/__init__.py:284]: why did you add this?

Comment (janhavlin) [reply tmt/steps/discover/__init__.py:284]: See @thrix's comment above. He asked me to add this error result.

File: tmt/steps/discover/__init__.py (hunk 4 above)
Comment (janhavlin) [reply tmt/steps/discover/__init__.py:10]: The `TYPE_CHECKING` variable? From my understanding, it helps with reducing the amount of imports. It is `True` only when doing type checking. When the code is being run in production, the imports in this `if` won't be imported as they are not needed - they're used only for typing.

Looking at the code, I'd say even the `from tmt.steps import StepData` line should be in this `if`.

Comment (guy9050) [reply tmt/steps/discover/__init__.py:10]: This looks like a good idea, I was asking since this condition is not applied in all the other modules where we have type annotation, so there's a consistency issue.. Do we want to add it to all the other modules? or are we okay that some have the conditional imports and some don't?
@thrix

File: tmt/steps/discover/__init__.py (hunk 5 above)
Comment (guy9050) [reply tmt/steps/discover/__init__.py:284]: +1

File: tmt/steps/discover/__init__.py (hunk 4 above)
Comment (thrix) [reply tmt/steps/discover/__init__.py:10]: It does not need to be in all modules, so it depends.

@janhavlin can you change `StepData` to `tmt.steps.Data` and remove it, seems it would bring some consistency here, or put it under and change other occurances to rather use `StepData`. I agree we should be consistent if possible.

Comment (thrix) [reply tmt/steps/discover/__init__.py:10]: Same seems to be applicable to `Test`. Or if you think it helps readability ... but to me more looks like inconsistency ...

**Review by thrix — CHANGES_REQUESTED** [review]

File: tmt/steps/discover/__init__.py [hunk 6]
```diff
@@ -234,9 +244,9 @@ def discover(context, **kwargs):
         return discover
 
     @classmethod
-    def options(cls, how=None):
+    def options(cls, how: Optional[str] = None) -> List[tmt.options.ClickOptionDecoratorType]:
         """ Prepare command line options for given method """
-        return [
+        return cast(List[tmt.options.ClickOptionDecoratorType], [
```
Comment (thrix) [inline tmt/steps/discover/__init__.py:249]: @janhavlin as discussed, this needs a change

File: tmt/steps/discover/__init__.py (hunk 4 above)
Comment (guy9050) [reply tmt/steps/discover/__init__.py:10]: Agreed

File: tmt/steps/discover/__init__.py (hunk 6 above)
Comment (janhavlin) [reply tmt/steps/discover/__init__.py]: Right, fixed.

**Comment by janhavlin** [comment]
Ok, @thrix, changed `StepData` to `tmt.steps.StepData` and `tmt.base.Test` to `tmt.Test` for consistency.

File: tmt/steps/execute/__init__.py [hunk 7]
```diff
@@ -245,7 +245,7 @@ def go(self, *args: Any, **kwargs: Any) -> None:
     def discover(self) -> tmt.steps.discover.Discover:
         """ Return discover plugin instance """
         # This is necessary so that upgrade plugin can inject a fake discover
-        return self.step.plan.discover
+        return cast(tmt.steps.discover.Discover, self.step.plan.discover)
```
Comment (janhavlin) [inline tmt/steps/execute/__init__.py:248]: Not sure why, but `tmt/steps/execute/__init__.py:248: error: Returning Any from function declared to return "Discover"` started to appear after I rebased to `main` (after merging #1327). If I remember correctly from annotating `gluetool-modules`, using `cast()` when `@property` decorator used is sometimes necessary due to some bug.

Comment (happz) [reply tmt/steps/execute/__init__.py:248]: You are correct, mymy should be able to infer the correct type. Please, add `cast()` and a big fat TODO item, this might be related to mypy not being allowed to follow imports - once we drop that restriction, `cast()` might become obsolete.

Comment (janhavlin) [reply tmt/steps/execute/__init__.py:252]: Right, added a TODO comment.

File: tmt/steps/__init__.py [hunk 8]
```diff
@@ -19,6 +19,7 @@
 from tmt.options import show_step_method_hints
 
 if TYPE_CHECKING:
+    import tmt
```
Comment (thrix) [inline tmt/steps/__init__.py:22]: @janhavlin this seems incorrect, `tmt` import is also a runtime dependency. See Build logs for errors which flap on this:
https://download.copr.fedorainfracloud.org/results/packit/teemtee-tmt-1333/epel-8-x86_64/04685763-tmt/builder-live.log.gz

**Review by thrix — CHANGES_REQUESTED** [review]

File: tmt/steps/__init__.py (hunk 8 above)
Comment (janhavlin) [reply tmt/steps/__init__.py]: Tried to import it so I can use `tmt.Test`. Ok, changed it to `from tmt.base import Test`.

**Review by thrix — APPROVED** [review]
LGTM

**Comment by thrix** [comment]
@janhavlin same problem with required runtime import

**Comment by janhavlin** [comment]
Not sure how to fix the errors. Can anyone help me with that please?

File: tmt/steps/__init__.py [hunk 9]
```diff
@@ -14,6 +14,8 @@
 import click
 from click import echo
 
+import tmt
+import tmt.base
```
Comment (happz) [inline tmt/steps/__init__.py:18]: I believe these two imports might be causing your issues.

```
__________________ ERROR collecting tests/unit/test_adjust.py __________________
tests/unit/test_adjust.py:3: in <module>
    from tmt.convert import relevancy_to_adjust
tmt/__init__.py:18: in <module>
    from tmt.base import Clean, Plan, Result, Run, Status, Story, Test, Tree
tmt/base.py:23: in <module>
    import tmt.steps
tmt/steps/__init__.py:73: in <module>
    class StepData(TypedDict, total=False):
tmt/steps/__init__.py:80: in StepData
    tests: Optional[List[tmt.base.Test]]
E   AttributeError: module 'tmt' has no attribute 'base'
```

`tmt` imports from `tmt.base`, which means `tmt.base` is read by the interpreter and processed. That leads to importing `tmt.steps` from `base.py`, then `steps.py` refers to `tmt.base.Test` - but `tmt.base` has not been fully imported yet, the process is still crunching `base.py` and `steps.py`, constructing the in-memory representation of those modules. This means `tmt` really has no `base` attribute, yet.

Try moving `import tmt.base` (is `import tmt` even needed?) under `TYPE_CHECKING` guard and a forward reference (`List['tmt.base.Test']`) to avoid referring to a partially imported module in runtime.

File: tmt/steps/discover/__init__.py [hunk 10]
```diff
@@ -1,24 +1,32 @@
...
 import tmt.utils
 
 
 class Discover(tmt.steps.Step):
     """ Gather information about test cases to be executed. """
 
-    def __init__(self, plan, data):
+    data: List[tmt.steps.StepData]
+
+    def __init__(self, plan: tmt.base.Plan, data: tmt.steps.StepData):
```
Comment (happz) [inline tmt/steps/discover/__init__.py:22]: These will be the same case, `tmt.base` is still being imported when `tmt.steps.discover` makes these references. Try forward references instead, i.e. `'tmt.base.Plan'`.

File: tmt/steps/__init__.py (hunk 9 above)
Comment (thrix) [reply tmt/steps/__init__.py]: seems resolved now, thanks @happz

File: tmt/steps/discover/__init__.py (hunk 10 above)
Comment (thrix) [reply tmt/steps/discover/__init__.py]: it is resolved

**Review by happz — APPROVED** [review]
